cou/steps/analyze.py

```diff
@@ -122,60 +104,49 @@ def __eq__(self, other: Any) -> Any:
 
-    def to_dict(self) -> Dict:
-        """Return a string in yaml format.
-
-        Passing the Application class directly to dump contain some fields that are big,
-        e.g: config and status. This output contains just the important fields for
-        the operator.
-        """
-        return {
+    def __str__(self) -> str:
+        """Dump as string."""
+        summary = {
             self.name: {
                 "model_name": self.model_name,
                 "charm": self.charm,
                 "charm_origin": self.charm_origin,
                 "os_origin": self.os_origin,
                 "channel": self.channel,
-                "pkg_name": self.pkg_name,
                 "units": {
                     unit: {
-                        "pkg_version": details.get("pkg_version", ""),
+                        "workload_version": details.get("workload_version", ""),
                         "os_version": details.get("os_version", ""),
                     }
                     for unit, details in self.units.items()
                 },
             }
         }
-
-    def __str__(self) -> str:
-        """Dump as string."""
-        return yaml.dump(self.to_dict())
-
-    def _get_pkg_name(self) -> str:
-        """Get the package name depending on the name of the charm."""
+        yaml = YAML()
+        stream = StringIO()
+        yaml.dump(summary, stream)
+        output_str = stream.getvalue()
+        stream.close()
+        return output_str
+
+    def _get_workload_name(self) -> str:
+        """Get the workload name depending on the name of the charm."""
         try:
-            pkg_name = CHARM_TYPES[self.charm]["pkg"]
+            workload = CHARM_TYPES[self.charm]["workload"]
         except KeyError:
-            logging.warning("package not found for application: %s", self.name)
-            pkg_name = ""
-        return pkg_name
+            logging.warning("workload not found for application: %s", self.name)
+            workload = ""
+        return workload
 
-    # NOTE (gabrielcocenza) Ideally, the application should provide the openstack version
-    # and packages versions by a charm action. This might be possible with Sunbeam.
-    async def _get_current_os_versions(self, unit: str) -> str:
+    def _get_current_os_versions(self, unit: str) -> str:
         """Get the openstack version of a unit."""
         version = ""
-        pkg_version = self._get_pkg_version(unit)
-        self.units[unit]["pkg_version"] = pkg_version
-        self.pkg_version_units[pkg_version].add(unit)
-
-        # for openstack releases >= wallaby
-        codename = await self._get_openstack_release(unit, model_name=self.model_name)
-        if codename:
-            version = codename
-        # for openstack releases < wallaby
-        elif self.pkg_name and pkg_version:
-            version = get_os_code_info(self.pkg_name, pkg_version)
+        workload_name = self._get_workload_name()
```

Comment:
the extra complexity comes from wanting to use `get_os_code_info` from zaza, which is meant to be used to determine codenames from on-disk packages, but actually feed it the workload_version that comes from the charm: so for example you'll come in with `charm=nova-compute, version=21.x.y` and need to remember that you need to convert nova-compute (the charm name) in the representative deb package name the function expects (nova-common). Eventually it would be better to extend `get_os_code_info` to accept both package or charm names.

Reply:
ultimately I would collapse `_get_workload_name` and `_get_charm_workload_version` into `_get_current_os_version`, they don't seem complex enough to warrant a separate function

Reply:
I think it is better to have them separate to ease of debugging in case of error.

Reply:
I agree with Erhan that it's better to have separate functions.

To try to solve this issue I've changed from `_get_workload_name` to `_get_representative_workload_pkg` and roll back the changes on `get_os_code_info`. I think now it's clear that we are taking the workload version of a charm, mapping to have a representative pkg and with that pass it to `get_os_code_info` and have the OpenStack code name.